rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,202 @@
+
+The `?VIEW_ROW_COUNT` is used to store the number of rows in the index and `?VIEW_KV_SIZE` keeps track of the size of this index. The size calculation is done using `erlang:external_size`.
+
+The process flow for a document to be indexed is as follows:
+
+1. FDB Transaction is started
+1. Read the document from the changes read (The number of documents to read at one type is configurable, the default is 100)
+1. The document is passed to the javascript query server and run through all the map functions defined in the design document
+1. The view's sequence number is updated to the sequence the document is in the changes feed.
+1. The emitted keys are stored in the `?VIEW_ID_RANGE`
+1. The emitted keys are encoded then added to the `?VIEW_MAP_RANGE` with the emitted keys and value stored
+1. The `?VIEW_ROW_COUNT` is incremented
+1. The `?VIEW_KV_SIZE` is increased
+1. If the document was deleted and was previously in the view, the previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE`. The Row count and size count are also decreased.
+1. If the document is being updated and was previously added to the index, then he previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE` and then the index is updated with the latest emitted keys and value.
+
+### Key ordering
+
+FoundationDB orders key by byte value which is not how CouchDB orders keys. To maintain CouchDB's view collation, a type value will need to be prepended to each key so that the correct sort order of null < boolean < numbers < strings < arrays < objects is maintained.
+
+In CouchDB 2.x, strings are compared via ICU. The way to do this with FoundationDB is that for every string an ICU sort string will be generated upfront and added to the key.
+
+### Emitting document
+
+In a map function, it is possible to emit the full document as the value, this will cause an issue if the document size is larger than FDB’s value limit of 100 KB. We can handle this in two possible ways. The first is to keep the hard limit of only allowing 100 KB value to be emitted, so if a document exceeds that CouchDB will return an error. This is the preferred option.
+
+The second option is to detect that a map function is emitting the full document and then add in a foreign key reference back to the document subspace. The issue here is that CouchDB would only be able to return the latest version of the document, which would cause consistency issues when combined with the `update=false` argument.
+
+A third option would be to split the document across multiple keys in FoundationDB. This will still be limited by the transaction size limit.
```

Comment:
Can this section be collapsed now into "the approach that has been chosen and why" and "the approaches that were not chosen and why".